File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/symmetric.ts
import {arrayToBTLevelOrder, BTNode} from "./BinaryTreeOperations";

function symmetric(left: BTNode, right: BTNode) {
	if (left === null || right === null) return left === right;
	if (left.data != right.data) return false;
	return symmetric(left.left, right.right) && symmetric(left.right, right.left);
}

function isSymmetric(root: BTNode) {
	return root === null || symmetric(root.left, root.right);
}

let root = arrayToBTLevelOrder([1,2,3,3,4,4,3]);
console.log(isSymmetric(root));

File: StriverA2Z/Java/BinaryTrees/Symmetric.java
public class Symmetric {
    public static void main(String[] args) {
        Node root = BinaryTreeOperations.arrayToBTLevelOrder(new Integer[]{1, 2, 2, 3, 4, 4, 3});
        System.out.print(isSymmetric(root));
    }

    public static boolean isSymmetric(Node root) {
        return root == null || isSymmetricHelp(root.left, root.right);
    }

    public static boolean isSymmetricHelp(Node left, Node right) {
        if (left == null || right == null) return left == right;
        if (left.data != right.data) return false;
        return isSymmetricHelp(left.left, right.right) && isSymmetricHelp(left.right, right.left);
    }
}
